### PR TITLE
Distinct sorted: calculate column positions once

### PR DESCRIPTION
**Pull request by devcrafter**
Calculate column positions once instead of calculating them for every chunk

Changelog category:
- Not for changelog (changelog entry is not required)

### CODE REVIEW

File: src/Processors/Transforms/DistinctSortedTransform.cpp [hunk 1]
```diff
@@ -119,24 +119,27 @@ bool DistinctSortedTransform::buildFilter(
...
 ColumnRawPtrs DistinctSortedTransform::getKeyColumns(const Chunk & chunk) const
 {
-    size_t columns = columns_names.empty() ? chunk.getNumColumns() : columns_names.size();
-
+    /// pre-calculate column positions to process during chunk transformation
+    if (column_positions.empty())
+    {
+        const size_t num_columns = column_names.empty() ? chunk.getNumColumns() : column_names.size();
+        column_positions.reserve(num_columns);
+        const Block& hdr = input.getHeader();
```
Comment (antonio2368) [inline src/Processors/Transforms/DistinctSortedTransform.cpp:127]:
```suggestion
        const Block & hdr = input.getHeader();
```

**Review by antonio2368 — APPROVED** [review]

File: src/Processors/Transforms/DistinctSortedTransform.h [hunk 2]
```diff
@@ -57,7 +58,8 @@ class DistinctSortedTransform : public ISimpleTransform
     };
     PreviousChunk prev_chunk;
 
-    Names columns_names;
+    Names column_names;
+    mutable ColumnNumbers column_positions;
```
Comment (alexey-milovidov) [inline src/Processors/Transforms/DistinctSortedTransform.h:62]: Why don't set it in the constructor from the header?

Comment (devcrafter) [reply src/Processors/Transforms/DistinctSortedTransform.h]: I did it this way at first (https://github.com/ClickHouse/ClickHouse/commit/416c14afd8c48483f118e3585a8cfe426f03e2ba, basically copy-paste from `DistinctTransform.cpp`), but a bit wrong and got confused by header with 0 columns.
Just recheck and see that `header_` is used in constructor, which is already moved into member`header`(unlike in `DistinctTransform`). So, the PR is back to calculation column positions in constructor.

**Comment by devcrafter** [comment]
@Mergifyio update

**Comment by Felixoid** [comment]
@Mergifyio update

**Pull request merged**